sympy/core/mul.py

```diff
@@ -1248,15 +1248,25 @@ def _eval_is_zero(self):
         return zero
 
     def _eval_is_integer(self):
+        from sympy import fraction
+
         is_rational = self.is_rational
+        if is_rational is False:
```

Comment:
Querying assumptions on self as part of an assumptions query also leads to problems. The infinite recursion is avoided because of this:
https://github.com/sympy/sympy/blob/a8eec2f93992cbee6eb666d9dc461ea30a56898b/sympy/core/assumptions.py#L296-L298
Instead it leads to non-deterministic results or cases where `is_integer` is `None` the first time you check it and then `True` the second time.

We should *try* to stick to only checking `is_*` properties on `args` rather than `self`.

A simple way to avoid it in this example would be to use `self._eval_is_rational` rather than `self.is_rational`.